### PR TITLE
Stats: Insights: Add annual site stats

### PR DESCRIPTION
**Pull request by psealock**
Add yearly stats as a stats module to Insights page

screen shot 2017-12-28 at 10 21 51 pm

cc @jancavan 

### Notes
* Average Comments is taken to one decimal place. Maybe this should only be for values < 10 or something?
* "View All" link at the bottom of the module can be included in the next PR that creates that page

### ToDo
- [x] Finalize design and update CSS and markup
- [x] i18n strings
- [x] Fix failing tests around the `utils` modification

### CODE REVIEW

**Comment by jancavan** [comment]
> should we have a fallback state if/when no data has been returned by the API?

Yes. We can go with "No annual stats recorded". Thoughts?

File: client/my-sites/stats/annual-site-stats/index.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'year' ) }</div>
+					<div className="annual-site-stats__stat-figure is-large">{ data.year }</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total posts' ) }</div>
+					<div className="annual-site-stats__stat-figure is-large">{ data.total_posts }</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total comments' ) }</div>
```
Comment (a8ci18n) [inline client/my-sites/stats/annual-site-stats/index.js]: Hi! I've found a possible matching string that has already been translated **35** times:
`translate( 'Total comments:' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by psealock** [comment]
### Update
* Tests now pass
* i18n strings
* Updated to match recent API changes
* Added "No annual stats recorded for this year." message

@jancavan FWIW here's the mobile view without collapsing any stats. Maybe its ok as is?

screen shot 2017-12-29 at 2 15 20 pm

File: client/my-sites/stats/annual-site-stats/index.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import StatsModulePlaceholder from 'my-sites/stats/stats-module/placeholder';
+import ErrorPanel from 'my-sites/stats/stats-error';
+
+class AnnualSiteStats extends Component {
+	static propTypes = {
+		requesting: PropTypes.bool,
+		siteId: PropTypes.number,
+		years: PropTypes.array,
+		translate: PropTypes.func,
+		moment: PropTypes.func,
```
Comment (timmyc) [inline client/my-sites/stats/annual-site-stats/index.js]: minor but missing `statType` here.

File: client/my-sites/stats/annual-site-stats/index.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		const { translate } = this.props;
+		return (
+			<div className="annual-site-stats__content">
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'year' ) }</div>
+					<div className="annual-site-stats__stat-figure is-large">{ data.year }</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total posts' ) }</div>
+					<div className="annual-site-stats__stat-figure is-large">{ data.total_posts }</div>
```
Comment (timmyc) [inline client/my-sites/stats/annual-site-stats/index.js]: All the numeric values should be passed through `this.props.numberFormat()` to localize the thousands separator and such.

File: client/my-sites/stats/annual-site-stats/index.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+	render() {
+		const { years, requesting, siteId, translate, moment, statType } = this.props;
+		const currentYear = moment().format( 'YYYY' );
+		const year = years && find( years, y => y.year === currentYear );
+		const isLoading = requesting && ! years;
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
+		return (
+			<div>
+				{ siteId && <QuerySiteStats siteId={ siteId } statType={ statType } /> }
```
Comment (timmyc) [inline client/my-sites/stats/annual-site-stats/index.js]: If we are only going to use this component in the context of the Insights page, we could reasonably rely upon the "Post Trends" component already making the query to the `statsInsights` endpoint. On this branch that endpoint is called 2x now on the page load with the same query params.

If we consider adding this component to the Years period page, then we would need to add the query component to ensure we have data to work with.

**Comment by psealock** [comment]
Thanks for the review @timmyc 
https://github.com/Automattic/wp-calypso/pull/21126/commits/5feb864f31fcc95b3ec264a431adf7de5ae9520e adds `numberFormat`, fixes propTypes, and removes extra query.

>On this branch that endpoint is called 2x now on the page load with the same query params.

Good catch. I've removed the query for now. We can add it back in if needed later.

>When loading the Insights page on an empty cache - the following is shown while the data is being fetched:

This is weird because the query component is rendered but doesn't immediately fire a request. I'm not sure why, but other components on the page suffer the same affliction because initially there is no data but `isRequesting === false`.

File: client/my-sites/stats/annual-site-stats/index.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+					<div className="annual-site-stats__stat-figure is-large">{ data.year }</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total posts' ) }</div>
+					<div className="annual-site-stats__stat-figure is-large">
+						{ numberFormat( data.total_posts ) }
+					</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total comments' ) }</div>
```
Comment (a8ci18n) [inline client/my-sites/stats/annual-site-stats/index.js]: Hi! I've found a possible matching string that has already been translated **35** times:
`translate( 'Total comments:' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by psealock** [comment]
https://github.com/Automattic/wp-calypso/pull/21126/commits/5f6dfb1cf5f89ad63e83212bad83b863aa811d7e determines loading state by presence of data instead of explicitly checking if the request is occurring. `noData` and `isError` states are also introduced.

@timmyc It should be more seamless while data is being fetched and/or waiting for the browser to fire a request.

File: client/my-sites/stats/annual-site-stats/index.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		const isLoading = ! years;
+		const isError = ! isLoading && years.errors;
+		const noData = ! isLoading && ! isError && ! year;
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
+		return (
+			<div>
+				<SectionHeader label={ translate( 'Annual Site Stats' ) } />
+				<Card className="stats-module">
+					<StatsModulePlaceholder isLoading={ isLoading } />
+					{ isError && <ErrorPanel message={ translate( 'Oops, something went wrong' ) } /> }
```
Comment (a8ci18n) [inline client/my-sites/stats/annual-site-stats/index.js]: Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Oops! Something went wrong.' )` *ES Score: 21*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/annual-site-stats)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by jancavan** [comment]
> should we consider moving it to the right column under Followers instead?

I think this module makes more sense to be next to "All time" so I think we should keep its located as is.

I've updated the CSS on mobile in: https://github.com/Automattic/wp-calypso/pull/21126/commits/39e2a9f6de0210ac6bf54b3ea38f408cb673532e

screenshot 2017-12-29 22 39 45

**Comment by beaulebens** [comment]
I noticed that now that we're into a new year, I get no results, and no way to view the previous year (which is actually the more interesting data!)

screen shot 2018-01-02 at 4 44 42 pm

**Comment by timmyc** [comment]
@beaulebens I _think_ the plan is to add in support for a "summary view" type page for this stat module which will allow viewing prior year's stats after this initial PR ships. It will be accessed via a "View All" link at the bottom of the module on the insights page:

view-all-0011

via p10gg3-8qN-p2

**Comment by blowery** [comment]
@beaulebens @timmyc Alternatively, maybe we show the previous year's stats while we're in January of the following year? Or show both for some range of time?

**Comment by blowery** [comment]
One possibility:

screenshot 2018-01-03 10 10 16

**Comment by blowery** [comment]
Added the second year while in january in https://github.com/Automattic/wp-calypso/pull/21126/commits/38235fb9655a832a355de2ea2f7f23bf0026f60b

File: client/my-sites/stats/annual-site-stats/index.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		const isLoading = ! years;
+		const isError = ! isLoading && years.errors;
+		const noData = ! isLoading && ! isError && ! ( currentYearData || previousYearData );
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
+		return (
+			<div>
+				<SectionHeader label={ translate( 'Annual Site Stats', { args: [ currentYear ] } ) } />
+				<Card className="stats-module">
+					<StatsModulePlaceholder isLoading={ isLoading } />
+					{ isError && <ErrorPanel message={ translate( 'Oops, something went wrong' ) } /> }
```
Comment (a8ci18n) [inline client/my-sites/stats/annual-site-stats/index.js]: Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Oops! Something went wrong.' )` *ES Score: 21*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/annual-site-stats)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by timmyc** [comment]
@psealock is out until the 10th so if we are feeling good about this @jancavan I think we should get it out there for stats fans.

**Comment by blowery** [comment]
@timmyc Jan is also out until the the 12th. I think we should just ship it.

**Comment by timmyc** [comment]
@blowery gotcha I think we can give it a merge too. One item though, with the addition of the second year, the mobile view looks like:

mobile-2-years

🤔  wondering if we should make the year row bold to help delineate the two

**Comment by blowery** [comment]
screenshot 2018-01-03 13 26 58

how's that?

**Comment by timmyc** [comment]
Looks good to me :shipit:

**Comment by psealock** [comment]
Thanks @jancavan @timmyc @blowery for adopting this one and seeing it through. Looking great

**Pull request merged**